actors/builtin/market/market_actor.go

```diff
@@ -181,94 +180,160 @@ func (a Actor) PublishStorageDeals(rt Runtime, params *PublishStorageDealsParams
+			rt.Log(rtt.INFO, "invalid deal %d: cannot publish deals from multiple providers in one batch", di)
+			continue
+		}
+		client, ok := rt.ResolveAddress(deal.Proposal.Client)
+		if !ok {
+			rt.Log(rtt.INFO, "invalid deal %d: failed to resolve proposal.Client address %v for deal ", di, deal.Proposal.Client)
+			continue
+		}
+
+		/*
+			drop deals with insufficient lock up to cover costs
+		*/
+		if _, ok := totalClientLockup[client]; !ok {
+			totalClientLockup[client] = abi.NewTokenAmount(0)
+		}
+		totalClientLockup[client] = big.Sum(totalClientLockup[client], deal.Proposal.ClientBalanceRequirement())
+		clientBalanceOk, err := msm.balanceCovered(client, totalClientLockup[client])
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to check client balance coverage")
+		if !clientBalanceOk {
+			rt.Log(rtt.INFO, "invalid deal: %d: insufficient client funds to cover proposal cost", di)
+			continue
+		}
+		totalProviderLockup = big.Sum(totalProviderLockup, deal.Proposal.ProviderCollateral)
+		providerBalanceOk, err := msm.balanceCovered(provider, totalProviderLockup)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to check provider balance coverage")
+		if !providerBalanceOk {
+			rt.Log(rtt.INFO, "invalid deal: %d: insufficient provider funds to cover proposal cost", di)
+			continue
+		}
+
+		/*
+			drop duplicate deals
+		*/
+		// Normalise provider and client addresses in the proposal stored on chain.
+		// Must happen after signature verification and before taking cid.
+		deal.Proposal.Provider = provider
+		resolvedAddrs[deal.Proposal.Client] = client
+		deal.Proposal.Client = client
+
+		pcid, err := deal.Proposal.Cid()
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalArgument, "failed to take cid of proposal %d", di)
+
+		// check proposalCids for duplication within message batch
+		// check state PendingProposals for duplication across messages
+		duplicateInState, err := msm.pendingDeals.Has(abi.CidKey(pcid))
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to check for existence of deal proposal")
+		_, duplicateInMessage := proposalCidLookup[pcid]
+		if duplicateInState || duplicateInMessage {
+			rt.Log(rtt.INFO, "invalid deal %d: cannot publish duplicate deal proposal %s", di)
+			continue
+		}
+
+		/*
+			check VerifiedClient allowed cap and deduct PieceSize from cap
+			drop deals with a DealSize that cannot be fully covered by VerifiedClient's available DataCap
+		*/
+		if deal.Proposal.VerifiedDeal {
+			code := rt.Send(
+				builtin.VerifiedRegistryActorAddr,
+				builtin.MethodsVerifiedRegistry.UseBytes,
+				&verifreg.UseBytesParams{
+					Address:  client,
+					DealSize: big.NewIntUnsigned(uint64(deal.Proposal.PieceSize)),
+				},
+				abi.NewTokenAmount(0),
+				&builtin.Discard{},
+			)
+			if code.IsError() {
+				rt.Log(rtt.INFO, "invalid deal %d: failed to acquire datacap exitcode: %d", code)
```

Comment:
I think you're missing the deal index in the log here

Reply:
nice thank you!